pkg/operator/csr/csr_approver.go

```diff
@@ -0,0 +1,275 @@
+package webhookcerts
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+
+	certapiv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/authentication/serviceaccount"
+	certv1informers "k8s.io/client-go/informers/certificates/v1"
+	certv1client "k8s.io/client-go/kubernetes/typed/certificates/v1"
+	certv1listers "k8s.io/client-go/listers/certificates/v1"
+	"k8s.io/klog/v2"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+type CSRApprovalDecision string
+
+const (
+	CSRApproved  CSRApprovalDecision = "Approved"
+	CSRDenied    CSRApprovalDecision = "Denied"
+	CSRNoOpinion CSRApprovalDecision = "NoOpinion"
+)
+
+type CSRApprover interface {
+	Approve(csrObj *certapiv1.CertificateSigningRequest, x509CSR *x509.CertificateRequest) (approvalStatus CSRApprovalDecision, denyReason string, err error)
+}
+
+type csrApproverController struct {
+	csrClient certv1client.CertificateSigningRequestInterface
+	csrLister certv1listers.CertificateSigningRequestLister
+
+	csrFilter   CSRFilter
+	csrApprover CSRApprover
+}
+
+// NewCSRApproverController returns a controller that is observing the CSR API
+// for a CSR of a given name. If such a CSR exists, it runs the `csrApprover.Approve()`
+// against it and either denies, approves or leaves the CSR.
+func NewCSRApproverController(
+	controllerName string,
+	operatorClient v1helpers.OperatorClient,
+	csrClient certv1client.CertificateSigningRequestInterface,
+	csrInformers certv1informers.CertificateSigningRequestInformer,
+	csrFilter CSRFilter,
+	csrApprover CSRApprover,
+	eventsRecorder events.Recorder,
+) factory.Controller {
+	c := &csrApproverController{
+		csrClient:   csrClient,
+		csrLister:   csrInformers.Lister(),
+		csrFilter:   csrFilter,
+		csrApprover: csrApprover,
+	}
+
+	return factory.New().
+		WithSync(c.sync).
+		WithSyncDegradedOnError(operatorClient).
+		WithInformers(csrInformers.Informer()).
+		ToController(
+			"WebhookAuthenticatorCertApprover_"+controllerName,
+			eventsRecorder.WithComponentSuffix("webhook-authenticator-cert-approver-"+controllerName),
+		)
+}
+
+func (c *csrApproverController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	csrList, err := c.csrLister.List(labels.Everything())
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	for _, csr := range csrList {
+		if !c.csrFilter.Matches(csr) {
```

Comment:
something passed in the night.  I'll merge this to have a starting point, but I think we want this filter function on every item we observe from the informer.

Reply:
make it a key function, I see. Can do in a follow-up, then.